aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,236 @@
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket()
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let nonce_key_with_exp_time = NonceKeyWithExpTime {
+            txn_expiration_time,
+            sender_address,
+            nonce,
+        };
+        let bucket_index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let current_time = timestamp::now_seconds();
+        if (!nonce_history.nonce_table.contains(bucket_index)) {
+            nonce_history.nonce_table.add(
+                bucket_index,
+                empty_bucket()
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, bucket_index);
+
+        let existing_exp_time = bucket.nonce_to_exp_time_map.get_cloned(&nonce_key);
+        if (existing_exp_time.is_some()) {
+            let existing_exp_time = existing_exp_time.extract();
+
+            // If the existing (address, nonce) pair has not expired, return false.
+            if (existing_exp_time >= current_time) {
+                return false;
+            };
+
+            // If the existing (address, nonce) pair has expired, garbage collect it.
+            bucket.nonce_to_exp_time_map.remove(&nonce_key);
+            bucket.nonces_ordered_by_exp_time.remove(&NonceKeyWithExpTime {
+                txn_expiration_time: existing_exp_time,
+                sender_address,
+                nonce,
+            });
+        };
+
+        // Garbage collect upto MAX_ENTRIES_GARBAGE_COLLECTED_PER_CALL expired nonces in the bucket.
+        let i = 0;
+        while (i < MAX_ENTRIES_GARBAGE_COLLECTED_PER_CALL) {
+            if (!bucket.nonces_ordered_by_exp_time.is_empty()) {
+                let (front_k, _) = bucket.nonces_ordered_by_exp_time.borrow_front();
+                if (front_k.txn_expiration_time < current_time) {
```

Comment:
same, should wait for MAX_EXPIRATION_TIME to expire?

Reply:
I mean if the user specified to expire the transaction in 30 seconds, why would we overwrite that expiration time with MAX_EXPIRATION_TIME seconds? So, I'm garbage collecting once the user specified expiration time passes.